rinex/src/qc/context.rs

```diff
@@ -1,61 +1,93 @@
-use crate::{Error, Rinex};
+use crate::Rinex;
 
 use sp3::prelude::SP3;
 
 #[derive(Default, Debug, Clone)]
-pub struct QcInputData {
-    /// File source path
+pub struct QcPrimaryData {
+    /// Source path
     pub path: PathBuf,
-    /// File Data
-    pub rinex: Rinex,
+    /// Data
+    pub data: Rinex,
 }
 
-impl QcInputData {
-    pub fn new(path: &str) -> Result<Self, Error> {
-        Ok(Self {
-            path: Path::new(path).to_path_buf(),
-            rinex: Rinex::from_file(path)?,
-        })
+#[derive(Default, Debug, Clone)]
+pub struct QcExtraData<T> {
+    /// Source paths
+    pub paths: Vec<PathBuf>,
+    /// Data
+    pub data: T,
+}
+
+impl<T> QcExtraData<T> {
+    /// Returns reference to Inner Data
+    pub fn data(&self) -> &T {
+        &self.data
+    }
+    /// Returns mutable reference to Inner Data
+    pub fn data_mut(&mut self) -> &mut T {
+        &mut self.data
+    }
+    /// Returns list of files that created this context
+    pub fn paths(&self) -> Vec<PathBuf> {
+        self.paths.clone()
     }
 }
 
 #[derive(Default, Debug, Clone)]
 pub struct QcContext {
-    /// Primary RINEX file
-    pub primary: QcInputData,
-    /// Optionnal NAV augmentation
-    pub nav: Option<QcInputData>,
-    /// Optionnal ATX data
-    pub atx: Option<QcInputData>,
-    /// Optionnal SP3 data
-    pub sp3: Option<SP3>,
-    /// SP3 files path
-    pub sp3_paths: Vec<PathBuf>,
+    /// Primary RINEX Data
+    pub primary: QcPrimaryData,
+    /// Optionnal NAV RINEX Data
+    pub nav: Option<QcExtraData<Rinex>>,
+    /// Optionnal ATX RINEX Data
+    pub atx: Option<QcExtraData<Rinex>>,
+    /// Optionnal SP3 Orbit Data
+    pub sp3: Option<QcExtraData<SP3>>,
 }
 
 impl QcContext {
-    /// Returns reference to primary data
-    pub fn primary_data(&self) -> &Rinex {
-        &self.primary.rinex
-    }
     /// Returns reference to primary data
     pub fn primary_path(&self) -> &PathBuf {
         &self.primary.path
     }
+    /// Returns reference to primary data
+    pub fn primary_data(&self) -> &Rinex {
+        &self.primary.data
+    }
+    /// Returns mutable reference to primary data
+    pub fn primary_data_mut(&mut self) -> &mut Rinex {
+        &mut self.primary.data
+    }
     /// Returns true if provided context contains
     /// navigation data, either as primary or subsidary data set.
     pub fn has_navigation_data(&self) -> bool {
-        self.primary.rinex.is_navigation_rinex() || self.nav.is_some()
+        self.primary.data.is_navigation_rinex() || self.nav.is_some()
+    }
+    /// Returns NAV files source path
+    pub fn nav_paths(&self) -> Vec<PathBuf> {
+        if let Some(ref nav) = self.nav {
+            nav.paths()
+        } else {
+            vec![]
+        }
     }
     /// Returns reference to navigation data specifically
     pub fn navigation_data(&self) -> Option<&Rinex> {
         if let Some(ref nav) = self.nav {
-            Some(&nav.rinex)
+            Some(&nav.data)
```

Comment:
`.map` might be more idiomatic, but doesn't matter.

Reply:
can't seem to find a more idiomatic way, that would not involve a Copy

Rinex does not implement copy. Currently it proposes something like this :

```rust 
pub fn navigation_data(&self) -> Option<&Rinex> {
      self.nav 
           .clone() // want to avoid that
           .map(|nav| nav.data)
           .as_ref()
}
```